### PR TITLE
Feat: add HTTPRoute-rule name to envoy route metadata

### LINKED ISSUE #4551: Add route-rule name to xDS route metadata
*Description*:
With the recent GatewayAPI V1.2 adoption, route rules are now named (https://github.com/kubernetes-sigs/gateway-api/pull/2985).

Currently when Envoy Gateway translates route rules to IR structs (and later to xDS resources), the name of the original HTTP route rule is not persisted to the route's metadata.

It is useful to have this information for runtime awareness of which GatewayAPI route was matched for each request (e.g. in access logging).

This can be achieved by setting the SectionName field on the route resource's metadata, which is currently left empty.

### PR DESCRIPTION
**Pull request by Ido-Itz**
This PR adds route rule names to the Envoy route metadata (section-name) as described in https://github.com/envoyproxy/gateway/issues/4551

This will allow various runtime flows (such as access logging) to have awareness of the matched route-rule.

Fixes #4551

Release Notes: No

### CODE REVIEW

**Comment by lsjostro** [comment]
would this mean I can target a http route rule section with a securitypolicy? 
like this?
```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: ext-auth-foo
spec:
  targetRefs:
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
      name: foobar
      sectionName: mainRule
  extAuth:
    grpc:
      backendRefs:
        - name: authserver
          namespace: authserver
          port: 8080

```

**Review by guydc — DISMISSED** [review]
LGTM

File: internal/gatewayapi/route.go [hunk 1]
```diff
@@ -362,6 +362,10 @@ func (t *Translator) processHTTPRouteRule(httpRoute *HTTPRouteContext, ruleIdx i
 			Name:               irRouteName(httpRoute, ruleIdx, matchIdx),
 			SessionPersistence: sessionPersistence,
 		}
+		irRoute.Metadata = buildRouteMetadata(httpRoute)
+		if rule.Name != nil {
+			irRoute.Metadata.SectionName = string(*rule.Name)
```
[the remark this answers is not in the document]
Comment (Ido-Itz) [reply internal/gatewayapi/route.go]: You're right. I missed the scenario of matcher-less routes :p

**Comment by Ido-Itz** [comment]
Fixed the issue of routes with no matchers + added support for gRPC routes as well

**Review by guydc — APPROVED** [review]

**Review by arkodg — APPROVED** [review]

**Pull request merged**